_posts/2021-06-21-a-new-tale.md

```diff
@@ -0,0 +1,43 @@
+---
+title: A new Tale - Pulp gets queueless distributed task workers
+author: Matthias Dellweg
+tags:
+  - tasking
+  - performance
+  - high availability
+  - scalability
+---
+
+Pulp is about to introduce a new replacement for its tasking system.
+This complete redesign comes with a much simpler architecture, that will drop the need for the resource-manager, the last singleton in the Pulp 3 infrastructure.
+Additionally, the distributed nature of the new task worker type drops the use of `Redis Queue` (`rq`).
+With this simplification there is no longer a need to replicate task data accross two databases.
+The main Postgres Pulp database will therefore be the single source of truth about tasks.
+Last but not least, the new worker type will use Postgres advisory locks to ensure exclusive access to tasks.
+This is a much more lightweight approach than storing access locks in a separate table.
+Also it simplifies the cleanup process vastly if a worker dies as postgres will clean the locks with its connection.
+
+With all these simplifications, the new tasking system is more reliable.
+Without the resource-manager as the single point of failure, it will be fully high available.
+Also with the resource-manager seen as the choke point in task processing, the new distributed system will scale much better in performance with the number of parallel running workers.
+
+> ![Tasking system benchmark](/images/tasking_system_benchmark/scale_workers.png)
+  This image shows the benchmark of the old and new tasking system with a single thread bursting 256 tasks that need a random 2 out of a pool of 128 abstract resources and need 0.5 s to complete.
+  Plotted is in the first panel the average time the foreground thread needs per task, and in the second the average time, a task needs from being available in the system and being picked up by a worker.
+  It can be seen, that in the old system, the `resource-manager` is able to deliver about 2 tasks per second, and that providing more than two workers do not provide additional speedup.
```

Comment:
```suggestion
  It can be seen in the old system that the `resource-manager` is able to deliver about 2 tasks per second, and that providing more than two workers do not provide additional speedup.
```